File: menu.py
#!/usr/bin/env python3
import smbus2
from adxl345 import ADXL345, OutputDataRate, Range  # adjust import if needed
import measure



def select_enum(enum_cls, default):
    print(
        f" Select {enum_cls.__name__} (press Enter for default: {default.name})"
    )

    members = list(enum_cls)
    for i, member in enumerate(members):
        print(f"{i}: {member.display}")


    choice = input("> ").strip()

    if choice == "":
        return default

    try:
        return members[int(choice)]
    except (ValueError, IndexError):
        print("Invalid selection. Using default.")
        return default


def confirm_settings(odr, g_range):
    print("\nSelected settings:")
    print(f"ODR   : {odr.name} ({odr.hz} Hz)")
    print(f"Range : {g_range.name}")

    confirm = input("Start measurement? (y/n): ").strip().lower()
    return confirm == "y"


def main():
    BUS_NUM = 1
    DEVICE_ADDR = 0x53  # typical ADXL345 address

    default_odr = OutputDataRate.ODR_100
    default_range = Range.RANGE_FULL

    odr = select_enum(OutputDataRate, default_odr)
    g_range = select_enum(Range, default_range)

    if not confirm_settings(odr, g_range):
        print("Startup aborted.")
        return

    bus = smbus2.SMBus(BUS_NUM)
    sensor = ADXL345(
        address=DEVICE_ADDR,
        bus=bus,
        odr=odr,
        g_range=g_range
    )

    measure.measure(sensor)


if __name__ == "__main__":
    main()

File: menu.cpp
/**
 * menu.cpp — interactive CLI menu for the ADXL345 C++ driver.
 *
 * Mirrors menu.py: lets the user choose OutputDataRate and Range,
 * confirm settings, then runs the measurement routine.
 *
 * Build:   make
 * Run:     sudo ./adxl345_measure
 *          (root or i2c group membership required for /dev/i2c-*)
 */

#include "adxl345.hpp"
#include "measure.hpp"

#include <iostream>
#include <string>
#include <vector>
#include <utility>
#include <stdexcept>

// ---------------------------------------------------------------------------
// select_odr — print a numbered list and return the user's choice.
// ---------------------------------------------------------------------------
static OutputDataRate select_odr(OutputDataRate default_odr) {
    const auto& list = odr_list();

    std::cout << "\n Select OutputDataRate"
              << " (press Enter for default: "
              << odr_to_name(default_odr) << ")\n";

    for (size_t i = 0; i < list.size(); ++i) {
        std::cout << i << ": " << list[i].second << '\n';
    }

    std::string line;
    std::cout << "> ";
    std::getline(std::cin, line);

    if (line.empty()) {
        return default_odr;
    }
    try {
        size_t idx = static_cast<size_t>(std::stoul(line));
        if (idx < list.size()) {
            return list[idx].first;
        }
    } catch (...) {}

    std::cout << "Invalid selection. Using default.\n";
    return default_odr;
}

// ---------------------------------------------------------------------------
// select_range — print a numbered list and return the user's choice.
// ---------------------------------------------------------------------------
static Range select_range(Range default_range) {
    const auto& list = range_list();

    std::cout << "\n Select Range"
              << " (press Enter for default: "
              << range_to_name(default_range) << ")\n";

    for (size_t i = 0; i < list.size(); ++i) {
        std::cout << i << ": " << list[i].second << '\n';
    }

    std::string line;
    std::cout << "> ";
    std::getline(std::cin, line);

    if (line.empty()) {
        return default_range;
    }
    try {
        size_t idx = static_cast<size_t>(std::stoul(line));
        if (idx < list.size()) {
            return list[idx].first;
        }
    } catch (...) {}

    std::cout << "Invalid selection. Using default.\n";
    return default_range;
}

// ---------------------------------------------------------------------------
// confirm_settings
// ---------------------------------------------------------------------------
static bool confirm_settings(OutputDataRate odr, Range g_range) {
    std::cout << "\nSelected settings:\n";
    std::cout << "ODR   : " << odr_to_name(odr)
              << " (" << odr_to_hz(odr) << " Hz)\n";
    std::cout << "Range : " << range_to_name(g_range) << '\n';

    std::cout << "Start measurement? (y/n): ";
    std::string line;
    std::getline(std::cin, line);
    return (!line.empty() && (line[0] == 'y' || line[0] == 'Y'));
}

// ---------------------------------------------------------------------------
// main
// ---------------------------------------------------------------------------
int main() {
    constexpr int     BUS_NUM     = 1;
    constexpr uint8_t DEVICE_ADDR = 0x1D;

    OutputDataRate default_odr   = OutputDataRate::ODR_100;
    Range          default_range = Range::RANGE_FULL;

    OutputDataRate odr     = select_odr(default_odr);
    Range          g_range = select_range(default_range);

    std::cout << "Stationary? (y/n): ";
    std::string stat_line;
    std::getline(std::cin, stat_line);
    bool is_stationary = (!stat_line.empty() &&
                          (stat_line[0] == 'y' || stat_line[0] == 'Y'));

    if (!confirm_settings(odr, g_range)) {
        std::cout << "Startup aborted.\n";
        return 0;
    }

    try {
        I2CBus  bus(BUS_NUM);
        ADXL345 sensor(DEVICE_ADDR, &bus, odr, g_range, /*watermark=*/16);
        measure(sensor, is_stationary);
    } catch (const std::exception& e) {
        std::cerr << "\nError: " << e.what() << '\n';
        return 1;
    }

    return 0;
}
